### PR TITLE
Update FV3atm/develop with changes from gmtb/develop as of 2019/10/16

### PR DESCRIPTION
**Pull request by climbfuji**
Associated PRs:

https://github.com/NCAR/ccpp-framework/pull/230
https://github.com/NCAR/ccpp-physics/pull/340
https://github.com/NOAA-EMC/GFDL_atmos_cubed_sphere/pull/5
https://github.com/NOAA-EMC/fv3atm/pull/1
https://github.com/ufs-community/ufs-weather-model/pull/1
https://github.com/NOAA-EMC/NEMS/pull/11 (merged)
https://github.com/NOAA-EMC/NEMS/pull/12

See https://github.com/ufs-community/ufs-weather-model/pull/1 for a description and regression testing information.

Changes specific to FV3atm:
- add ccpp-framework and ccpp-physics submodules as `ccpp/framework` and `ccpp/physics`
- move contents of `ufs-weather-model/ccpp` to `ccpp`
move contents of `CCPP_layer` to `ccpp/driver`, adjust `makefile` and `CMakeLists.txt`
- update CCPP suite definition files and `gfsphysics/GFS_layer/GFS_typedefs.*` to reflect GFSv16 changes in CCPP
- bugfixes for IPD physics as described in https://vlab.ncep.noaa.gov/redmine/issues/69917
- remove `TRANSITION` mode
- stub inline post module for systems that do not have EMC-post installed

### CODE REVIEW

File: gfsphysics/physics/GFS_debug.F90
Comment (junwang-noaa) [inline gfsphysics/physics/GFS_debug.F90]: fix typo: CCPP only

Comment (climbfuji) [reply gfsphysics/physics/GFS_debug.F90]: Thanks, fixed it!

File: io/FV3GFS_io.F90
Comment (junwang-noaa) [inline io/FV3GFS_io.F90]: why do we need this?

File: io/FV3GFS_io.F90
Comment (junwang-noaa) [inline io/FV3GFS_io.F90]: Can we remove this #if 1?

File: .gitmodules
Comment (junwang-noaa) [inline .gitmodules]: So this will be change back to EMC fork after fv3 dycore is committed?

**Review by junwang-noaa — COMMENTED** [review]
The code looks good to me, please see the comments in the code, fix should be minor.

File: io/FV3GFS_io.F90
Comment (climbfuji) [reply io/FV3GFS_io.F90]: This is not really needed. It's more a reminder that we can remove this once the Noah-MP initialization/calculation of variables in this file is moved to CCPP. We do have the code ready inside CCPP, but the problem is that the calculation in FV3GFS_io.F90 follows the precision of the FV3 code outside of gfsphysics, i.e. it is 32bit in case the code is built with 32BIT=Y, while the calculation in CCPP is always in the physics precision (64bit). When the IPD is phased out, we need to agree on doing this calculation in CCPP in 64bit (or whatever the precision for the physics is). Long story short, I will remove those #if 1 and corresponding #endif statements.

File: .gitmodules
Comment (climbfuji) [reply .gitmodules]: Yes, this is what Dusan suggested in his email. Once the code is committed to https://github.com/NOAA-EMC/GFDL_atmos_cubed_sphere, we can check out the head of https://github.com/NOAA-EMC/GFDL_atmos_cubed_sphere's develop branch, update the submodule pointer in fv3atm to point to this hash, and reset the .gitmodules file.

File: io/FV3GFS_io.F90
Comment (climbfuji) [reply io/FV3GFS_io.F90]: I updated this file, please have a look. Thanks!

**Review by DusanJovic-NOAA — APPROVED** [review]

File: gfsphysics/GFS_layer/GFS_typedefs.F90
Comment (SMoorthi-emc) [inline gfsphysics/GFS_layer/GFS_typedefs.F90:745]: Where does the UGW in GFSv16 fall under this category?

**Review by SMoorthi-emc — APPROVED** [review]
Looks OK to me

File: gfsphysics/GFS_layer/GFS_typedefs.F90
Comment (climbfuji) [reply gfsphysics/GFS_layer/GFS_typedefs.F90:745]: The UGW and original GFS gravity wave drag are controlled as before by their own variables (cnvgwd, do_tofd, ...) - this gwd_opt is only effective when the GSD drag suite is in the suite definition file. Once the GFS-UGW gravity wave drag code has converged to a clean solution (the current way of calling gwdps from inside UGWP is a little confusing), we should straighten out all the options and control switches in my opinion.

**Review by climbfuji — COMMENTED** [review]
Moorthi, please see my reply to your comment further down.

**Comment by SMoorthi-emc** [comment]
Thanks.
Frankly, whole CCPP is one too complex system.
Moorthi

**Comment by climbfuji** [comment]
> Thanks. Frankly, whole CCPP is one too complex system. Moorthi
> […](#)
> Sent from my iPhone
> On Oct 24, 2019, at 6:15 PM, Dom Heinzeller ***@***.***> wrote: @climbfuji commented on this pull request. In gfsphysics/GFS_layer/GFS_typedefs.F90: > @@ -744,6 +741,12 @@ real (kind=kind_phys), pointer :: prsik (:,:) => null() !< Exner function at i logical :: trans_trac !< flag for convective transport of tracers (RAS, CS, or SAMF) logical :: old_monin !< flag for diff monin schemes logical :: cnvgwd !< flag for conv gravity wave drag +#ifdef CCPP + integer :: gwd_opt !< gwd_opt = 1 => original GFS gwd (gwdps.f) The UGW and original GFS gravity wave drag are controlled as before by their own variables (cnvgwd, do_tofd, ...) - this gwd_opt is only effective when the GSD drag suite is in the suite definition file. Once the GFS-UGW gravity wave drag code has converged to a clean solution (the current way of calling gwdps from inside UGWP is a little confusing), we should straighten out all the options and control switches in my opinion. — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub, or unsubscribe.

I appreciate your feedback, but I am not sure if this is the right place for it.

File: io/FV3GFS_io.F90
Comment (SMoorthi-emc) [reply io/FV3GFS_io.F90]: Please do not commit this code yet until it is merged with fv3atm master.
I just found out the GFS_typedef has old code in there such as "hprim", "lgocart" etc.
IPlease verify that it is merged before committing.
Thanks
Moorthi

Comment (climbfuji) [reply io/FV3GFS_io.F90]: Hi Moorthi, they code is 100% up to date with the latest master.

Can you please clarify where hprim is in GFS_typedefs.F90? I searched the file, the only one used is Sfprop%hprime. The interstitial variable Interstitial%hprime1 has been removed completely. The one definition that was left mistakenly is in lines 1042-1043 (inside an `#ifdef CCPP` block):
```
    ! Temporary to support transition to GFSv16 physics
    logical :: lgocart = .false.
```
but since it is not used I can remove it (or leave the cleanup for another commit if Jun/Dusan do not want to change the code anymore for regression testing):
```
imk-ifu-l257b1:ufs-weather-model_update_from_gmtb-develop_20191016 heinzeller-d$ grep -Rie 'lgocart' FV3/gfsphysics/GFS_layer/GFS_typedefs.* FV3/ccpp
FV3/gfsphysics/GFS_layer/GFS_typedefs.F90:    logical :: lgocart = .false.
FV3/gfsphysics/GFS_layer/GFS_typedefs.meta:[lgocart]
```

Comment (climbfuji) [reply io/FV3GFS_io.F90]: @junwang-noaa @DusanJovic-NOAA please let me know if you want me to remove lines 1042-1043 (depending on how far your regression tests got). Thanks! And thanks @SMoorthi-emc for spotting these two lines.

**Comment by SMoorthi-emc** [comment]
I first found that issue in the file you recommended for me to copy for ras implementation. I found that version was old.
Then I went on github and I must have seen a wrong branch. Later I and Jun looked together it seems ok. I just wanted folks to double check as I am not too comfortable checking code on github.
Nevertheless, it appears there is no issue although the code you recommended to me is behind.
Moorthi

**Comment by climbfuji** [comment]
Hi Moorthi,

the CCPP annotations are not yet in the EMC authoritative repositories, this will happen in the near future I hope. The version I pointed you to must have been the NCAR gmtb/develop branch before the GFSv16 changes were merged in from NEMSfv3gfs/master.

Are you aware that you can change the diff view in GitHub to resemble more the gerrit look? See screenshot, change from unified to split.

Dom


Screen Shot 2019-10-26 at 5 24 13 PM

**Comment by climbfuji** [comment]
I removed the leftover CCPP-only lgocart definition in GFS_typedefs.F90 and GFS_typedefs.meta. Code compiles and runs fine (it wasn't used anywhere in CCPP anymore).

File: io/FV3GFS_io.F90
Comment (junwang-noaa) [reply io/FV3GFS_io.F90]: I would suggest to leave it to the next commit.  I added two code changes to my branch for the bug fixes prepared for next commit

**Review by junwang-noaa — APPROVED** [review]
The code changes look good to me

**Pull request merged**
